src/agent/internals/src/main/java/com/google/devtools/cdbg/debuglets/java/GcpEnvironment.java

```diff
@@ -173,6 +175,36 @@ static Map<String, String> getDebuggeeLabels() {
     return labels;
   }
 
+  /**
+   * Returns the user dir of a Java8 App Engine Standard instance. This is the base directory where
+   * the user's files are deployed. If no Java 8 App Engine Standard user dir could be determined
+   * null is returned.  It is safe to call this regardless of environment, as long as the return
+   * value is checked.
+   *
+   * <p> While there is a system property defined, "user.dir", in Java8 App Engine environments, it
+   * is unfortunately not set at the point in time this agent code runs, so it cannot be relied
+   * upon. Here we use the heuristic that the WEB-INF directory is deployed at the root of the user
+   * dir, and the deployed agent is deployed somewhere under the user dir. This means we can search
+   * through the agent directory's ancestors looking for the WEB-INFO directory, and once found
+   * we've found the user dir.
+   *
+   * @return the Java 8 App Engine Standard user directory if it was able to be determined, null
+   * otherwise.
+   */
+  public static String getAppEngineJava8UserDir(String agentDir) {
```

Comment:
Sad that we can't use Optional for the return type.

Probably rename to "tryGetAppEngineJava8UserDir" or similar to emphasize that the attempt is allowed to fail gracefully.

Reply:
Done